dlt/common/destination/reference.py

```diff
@@ -344,59 +345,140 @@ def should_truncate_table_before_load_on_staging_destination(self, table: TTable
+
+#     @staticmethod
+#     def from_name(destination: TDestinationReferenceArg) -> "DestinationReference":
+#         if destination is None:
+#             return None
+
+#         # if destination is a str, get destination reference by dynamically importing module
+#         if isinstance(destination, str):
+#             try:
+#                 if "." in destination:
+#                     # this is full module name
+#                     destination_ref = cast(DestinationReference, import_module(destination))
+#                 else:
+#                     # from known location
+#                     destination_ref = cast(DestinationReference, import_module(f"dlt.destinations.impl.{destination}"))
+#             except ImportError:
+#                 if "." in destination:
+#                     raise UnknownDestinationModule(destination)
+#                 else:
+#                     # allow local external module imported without dot
+#                     try:
+#                         destination_ref = cast(DestinationReference, import_module(destination))
+#                     except ImportError:
+#                         raise UnknownDestinationModule(destination)
+#         else:
+#             destination_ref = cast(DestinationReference, destination)
+
+#         # make sure the reference is correct
+#         try:
+#             c = destination_ref.spec()
+#             c.credentials
+#         except Exception:
+#             raise InvalidDestinationReference(destination)
+
+#         return destination_ref
+
+#     @staticmethod
+#     def to_name(destination: TDestinationReferenceArg) -> str:
+#         if isinstance(destination, DestinationFactory):
+#             return destination.name
+#         if isinstance(destination, ModuleType):
+#             return get_module_name(destination)
+#         return destination.split(".")[-1]  # type: ignore
+
+
+class Destination(ABC):
+    """A destination factory that can be partially pre-configured
+    with credentials and other config params.
+    """
+    credentials: Optional[CredentialsConfiguration] = None
+    config_params: Optional[Dict[str, Any]] = None
 
+    def __init__(self, cfg: DestinationClientConfiguration) -> None:
+        cfg_dict = dict(cfg)
+        self.credentials = cfg_dict.pop("credentials", None)
+        self.config_params = {key: val for key, val in cfg_dict.items() if val is not None}
 
-class DestinationReference(Protocol):
-    __name__: str
-    """Name of the destination"""
+    @property
+    @abstractmethod
+    def spec(self) -> Type[DestinationClientConfiguration]:
+        """Returns the destination configuration spec"""
+        ...
 
+    @property
+    @abstractmethod
     def capabilities(self) -> DestinationCapabilitiesContext:
-        """Destination capabilities ie. supported loader file formats, identifier name lengths, naming conventions, escape function etc."""
+        """Returns the destination capabilities"""
```

Comment:
this should always return a new instance. there's a lot of code that takes capabilities and then modifies the instance ie. to restrict the file formats or change the preferred format. so I'd say make it explicit method makes it more obvious.